extension/popup.js

```diff
@@ -45,23 +48,87 @@ const utils = {
             height: 630,
             type: 'popup'
         };
+        return chrome.windows.create({ ...defaultOptions, ...options, url });
+    },
+
+    async fetchYouTubeTranscript(videoId) {
+        try {
+            const response = await fetch(`https://www.youtube.com/watch?v=${videoId}`, {
+                headers: {
+                    'Accept-Language': 'en-US',
+                    'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36'
+                }
+            });
+
+            const html = await response.text();
+            const splittedHtml = html.split('"captions":');
+
+            if (splittedHtml.length <= 1) {
+                throw new Error('No captions found in video');
+            }
+
+            const captionsJson = JSON.parse(
+                splittedHtml[1].split(',"videoDetails')[0].replace(/\n/g, '')
+            );
+
+            const captionTracks = captionsJson?.playerCaptionsTracklistRenderer?.captionTracks;
+            if (!captionTracks?.length) {
+                throw new Error('No caption tracks found');
+            }
 
-        const windowOptions = { ...defaultOptions, ...options, url };
-        return chrome.windows.create(windowOptions);
+            const track = captionTracks.find(t => t.languageCode === 'en') || captionTracks[0];
+            const transcriptResponse = await fetch(track.baseUrl);
+            const transcriptXml = await transcriptResponse.text();
+
+            return this.parseTranscriptXml(transcriptXml);
+        } catch (error) {
+            console.error('Error fetching transcript:', error);
+            throw new Error(`Failed to fetch transcript: ${error.message}`);
+        }
+    },
```

Comment:
This is all for a browser extension. Definitely appreciate the suggestions and normally this would apply however the point is to use scraping since we're doing this within an extension.